staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cbor
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+var em cbor.EncMode = func() cbor.EncMode {
+	em, err := cbor.EncOptions{
+		Sort:          cbor.SortBytewiseLexical,
+		ShortestFloat: cbor.ShortestFloat16,
+		NaNConvert:    cbor.NaNConvert7e00,
+		InfConvert:    cbor.InfConvertFloat16,
+		BigIntConvert: cbor.BigIntConvertShortest,
+		Time:          cbor.TimeRFC3339Nano,
+		TimeTag:       cbor.EncTagNone,
+		IndefLength:   cbor.IndefLengthForbidden,
+		NilContainers: cbor.NilContainerAsNull,
+		TagsMd:        cbor.TagsAllowed,
+	}.EncMode()
+	if err != nil {
+		panic(err)
+	}
+	return em
+}()
+
+var dm cbor.DecMode = func() cbor.DecMode {
+	dm, err := cbor.DecOptions{
+		DupMapKey:         cbor.DupMapKeyEnforcedAPF,
+		TimeTag:           cbor.DecTagRequired,
+		MaxNestedLevels:   0,
+		MaxArrayElements:  0,
+		MaxMapPairs:       0,
+		IndefLength:       cbor.IndefLengthForbidden,
+		TagsMd:            cbor.TagsAllowed,
+		IntDec:            cbor.IntDecConvertSigned,
+		MapKeyByteString:  cbor.MapKeyByteStringForbidden,
+		ExtraReturnErrors: cbor.ExtraDecErrorUnknownField,
+		DefaultMapType:    reflect.TypeOf(map[string]interface{}(nil)),
+		UTF8:              cbor.UTF8RejectInvalid,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+	return dm
+}()
+
+type metaFactory interface {
+	// Interpret should return the version and kind of the wire-format of the object.
+	Interpret(data []byte) (*schema.GroupVersionKind, error)
+}
+
+type defaultMetaFactory struct{}
+
+var mfdm cbor.DecMode = func() cbor.DecMode {
```

Comment:
Why should we have a different set of decoding options for pulling metadata?  I'd expect a single set.

Any response should be a comment here because I bet future me will wonder the same thing.